% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/set_vars.R
\name{set_vars}
\alias{set_vars}
\title{Function to process and set the arguments passed by the user for each function}
\usage{
set_vars(
  roi,
  bam_file,
  genome,
  path_to_RNAfold,
  path_to_RNAplot,
  plot_output = TRUE,
  pi_pal = c("RdYlBl", "BlYel", "yelOrRed", "MagYel", "Greens"),
  si_pal = c("RdYlBl", "BlYel", "yelOrRed", "MagYel", "Greens"),
  weight_reads = c("None", "top", "locus_norm", "none", "Top", "Locus_Norm"),
  write_fastas = FALSE,
  annotate_region = FALSE,
  gtf_file = FALSE,
  out_type = c("pdf", "png", "PDF", "PNG"),
  use_bed_names = FALSE
)
}
\arguments{
\item{roi}{The path to a BED file of loci of interest}

\item{bam_file}{The path to a BAM file}

\item{genome}{The path to a genome Fasta file}

\item{path_to_RNAfold}{The full path to the RNAfold binary executable.}

\item{path_to_RNAplot}{The full path to the RNAplot binary executable.}

\item{plot_output}{Determines whether the program will output plots as PDFs.
Expected input is TRUE or FALSE.}

\item{pi_pal}{The color palette to use for the piRNA heatmap plot.
Valid options are "RdYlBl", "BlYel", "yelOrRed", "MagYel", and "Greens".}

\item{si_pal}{The color palette to use for the siRNA heatmap plot.
Valid options are "RdYlBl", "BlYel", "yelOrRed", "MagYel", and "Greens".}

\item{weight_reads}{Determines whether read counts will be weighted.
Valid options are "Top", "locus_norm", or "None".
See MiSiPi documentation for descriptions of the weighting methods.}

\item{write_fastas}{TRUE or FALSE. Optional.
If TRUE, read pairs from functions will be written to file.}

\item{annotate_region}{Determines whether the program will plot genomic
features of interest found in the GTF annotation file.
If TRUE, a GTF file must be provided as the "gtf_file" argument.}

\item{gtf_file}{a string corresponding to the path of genome
annotation in 9-column GTF format.
Default is FALSE unless annotate_regions == TRUE.}

\item{out_type}{The type of file for plots.
Options are "png" or "pdf". Default is PDF.}

\item{use_bed_names}{a boolean indicating if the name column in
the bed file should be used for results.
If TRUE, bed line names will be used unless column 4 is not present in the bed file.
In that case, this parameter will toggle to FALSE.
If FALSE, results will be named using a region string in the format: chr-start_stop}
}
\value{
a list
}
\description{
Function to process and set the arguments passed by the user for each function
}
